File: Leetcode/hashing/existing/find-players-with-zero-or-one-losses/findPlayersWithZeroOrOneLosses.ts
function findWinners(matches: number[][]): number[][] {
    const zeroLoss = new Set<number>();
    const oneLoss = new Set<number>();
    const moreLoss = new Set<number>();

    for (const [winner, loser] of matches) {
        if (!oneLoss.has(winner) && !moreLoss.has(winner)) {
            zeroLoss.add(winner);
        }
        if (zeroLoss.has(loser)) {
            zeroLoss.delete(loser);
            oneLoss.add(loser);
        } else if (oneLoss.has(loser)) {
            oneLoss.delete(loser);
            moreLoss.add(loser);
        } else if (moreLoss.has(loser)) {
            continue;
        } else {
            oneLoss.add(loser);
        }
    }

    return [
        Array.from(zeroLoss).sort((a, b) => a - b),
        Array.from(oneLoss).sort((a, b) => a - b)
    ];
}

// Time: O(n * log n);
// Space: O(n);

const result = findWinners([[1, 3], [2, 3], [3, 6], [5, 6], [5, 7], [4, 5], [4, 8], [4, 9], [10, 4], [10, 9]]);
console.log(result);

File: Leetcode/hashing/existing/find-players-with-zero-or-one-losses/findPlayersWithZeroOrOneLosses.java
public List<List<Integer>> findWinners(int[][] matches) {
    Set<Integer> seen = new HashSet<>();
    Map<Integer, Integer> losersCount = new HashMap<>();

    for (int[] match : matches) {
        int winner = match[0];
        int loser = match[1];
        seen.add(winner);
        seen.add(loser);

        losersCount.put(loser, losersCount.getOrDefault(loser, 0) + 1);
    }

    List<List<Integer>> answer = Arrays.asList(new ArrayList<>(), new ArrayList<>());

    for (int player : seen) {
        if (!losersCount.containsKey(player)) {
            answer.get(0).add(player);
        } else if (losersCount.get(player) == 1) {
            answer.get(1).add(player);
        }
    }

    Collections.sort(answer.get(0));
    Collections.sort(answer.get(1));

    return answer;
}

// Time: O(n * log n);
// Space: O(n);
